Domain/Academy/src/main/java/tw/waterball/judgegirl/academy/domain/usecases/exam/CalculateExamScoreUseCase.java

```diff
@@ -0,0 +1,105 @@
+import tw.waterball.judgegirl.commons.utils.functional.GetById;
+import tw.waterball.judgegirl.primitives.Student;
+import tw.waterball.judgegirl.primitives.exam.Exam;
+import tw.waterball.judgegirl.primitives.exam.Examinee;
+import tw.waterball.judgegirl.primitives.exam.Question;
+import tw.waterball.judgegirl.primitives.exam.Record;
+import tw.waterball.judgegirl.studentapi.clients.StudentServiceDriver;
+
+import javax.inject.Named;
+import java.util.List;
+
+import static java.util.function.Function.identity;
+import static tw.waterball.judgegirl.commons.utils.StreamUtils.*;
+
+/**
+ * @author - johnny850807@gmail.com (Waterball)
+ */
+@Named
+public class CalculateExamScoreUseCase extends AbstractExamUseCase {
+    private final StudentServiceDriver studentServiceDriver;
+
+    public CalculateExamScoreUseCase(ExamRepository examRepository, StudentServiceDriver studentServiceDriver) {
+        super(examRepository);
+        this.studentServiceDriver = studentServiceDriver;
+    }
+
+    public void execute(int examId, Presenter presenter) {
+        Exam exam = findExam(examId);
+        List<Record> records = examRepository.findAllRecordsInAnExam(examId);
+
+        var examinees = examinees(exam);
+        List<ExamineeRecord> examineeRecords = examineeRecords(exam, records, examinees);
+        double averageScore = average(examineeRecords, ExamineeRecord::getTotalScore);
+        present(presenter, exam, examineeRecords, averageScore);
+    }
+
+    private List<Record> filterRecords(List<Record> records, GetById<Integer, Student> examinees) {
+        return filterToList(records, record -> examinees.get(record.getStudentId()) != null);
+    }
+
+    private GetById<Integer, Student> examinees(Exam exam) {
+        var idToExaminee = toMap(studentServiceDriver.getStudentsByIds(
+                mapToList(exam.getExaminees(), Examinee::getStudentId)), Student::getId, identity());
+        return idToExaminee::get;
+    }
+
+    private List<ExamineeRecord> examineeRecords(Exam exam, List<Record> records, GetById<Integer, Student> examinees) {
+        var examineeToQuestionRecords =
+                groupingBy(questionRecords(exam, records),
+                        questionRecord -> examinees.get(questionRecord.getStudentId()));
+        return zipToList(examineeToQuestionRecords, ExamineeRecord::new);
+    }
+
+    private List<QuestionRecord> questionRecords(Exam exam, List<Record> records) {
+        return mapToList(records, record -> new QuestionRecord(exam.getQuestionById(record.getQuestionId()), record));
+    }
+
+    private void present(Presenter presenter, Exam exam, List<ExamineeRecord> examineeRecords, double averageScore) {
+        presenter.showExam(exam);
+        examineeRecords.forEach(presenter::addRecord);
+        presenter.showStatistics(averageScore, exam.getMaxScore());
+    }
+
+    public interface Presenter {
+
+        void showExam(Exam exam);
+
+        void addRecord(ExamineeRecord examineeRecord);
```

Comment:
@Wally5077 Note, this is the new naming convention (`addXXX`), when we are going to present multiple elements.